pgmax/fg/graph.py

```diff
@@ -45,181 +43,86 @@ class FactorGraph:
-            To add an ORFactor to a FactorGraph fg, run::
-
-                fg.add_factor_by_type(
-                    variable_names=variables_names_for_OR_factor,
-                    factor_type=logical.ORFactor
-                )
-        """
-        if factor_type not in FAC_TO_VAR_UPDATES:
-            raise ValueError(
-                f"Type {factor_type} is not one of the supported factor types {FAC_TO_VAR_UPDATES.keys()}"
-            )
-
-        name = kwargs.pop("name", None)
-        variables = tuple(self._variable_group[variable_names])
-        factor = factor_type(variables, *args, **kwargs)
-        factor_group = groups.SingleFactorGroup(
-            variable_group=self._variable_group,
-            variable_names_for_factors=[variable_names],
-            factor=factor,
-        )
-        self._register_factor_group(factor_group, name)
-
-    def add_factor_group(self, factory: Callable, *args, **kwargs) -> None:
-        """Add a factor group to the factor graph
-
-        Args:
-            factory: Factory function that takes args and kwargs as input and outputs a factor group.
-            args: Args to be passed to the factory function.
-            kwargs: kwargs to be passed to the factory function, and an optional "name" argument
-                for specifying the name of a named factor group.
+        Raises:
+            ValueError: If
+                (1) Both a Factor and a FactorGroup are added
+                (2) The FactorGroup involving the same variables already exists in the FactorGraph.
         """
-        name = kwargs.pop("name", None)
-        factor_group = factory(self._variable_group, *args, **kwargs)
-        self._register_factor_group(factor_group, name)
-
-    def _register_factor_group(
-        self, factor_group: groups.FactorGroup, name: Optional[str] = None
-    ) -> None:
-        """Register a factor group to the factor graph, by updating the factor graph state.
+        if factor is None and factor_group is None:
+            raise ValueError("A Factor or a FactorGroup is required")
 
-        Args:
-            factor_group: The factor group to be registered to the factor graph.
-            name: Optional name of the factor group.
+        if factor is not None and factor_group is not None:
+            raise ValueError("Cannot simultaneously add a Factor and a FactorGroup")
 
-        Raises:
-            ValueError: If the factor group with the same name or a factor involving the same variables
-                already exists in the factor graph.
-        """
-        if name in self._named_factor_groups:
-            raise ValueError(
-                f"A factor group with the name {name} already exists. Please choose a different name!"
+        if factor is not None:
+            factor_group = groups.SingleFactorGroup(
+                variables_for_factors=[factor.variables],
+                factor=factor,
             )
+        assert factor_group is not None
```

Comment:
Let's not do assert here but instead make sure the logic is solid and `factor_group` can never be `None`